libraries/chain/asset_evaluator.cpp

```diff
@@ -360,42 +360,127 @@ void_result asset_update_issuer_evaluator::do_apply(const asset_update_issuer_op
    return void_result();
 } FC_CAPTURE_AND_RETHROW( (o) ) }
 
-void_result asset_update_bitasset_evaluator::do_evaluate(const asset_update_bitasset_operation& o)
+void_result asset_update_bitasset_evaluator::do_evaluate(const asset_update_bitasset_operation& op)
 { try {
    database& d = db();
 
-   const asset_object& a = o.asset_to_update(d);
+   const asset_object& asset_obj = op.asset_to_update(d);
 
-   FC_ASSERT(a.is_market_issued(), "Cannot update BitAsset-specific settings on a non-BitAsset.");
+   FC_ASSERT( asset_obj.is_market_issued(), "Cannot update BitAsset-specific settings on a non-BitAsset." );
 
-   const asset_bitasset_data_object& b = a.bitasset_data(d);
-   FC_ASSERT( !b.has_settlement(), "Cannot update a bitasset after a settlement has executed" );
-   if( o.new_options.short_backing_asset != b.options.short_backing_asset )
+   FC_ASSERT( op.issuer == asset_obj.issuer, "Only asset issuer can update bitasset_data of the asset." );
+
+   const asset_bitasset_data_object& current_bitasset_data = asset_obj.bitasset_data(d);
+
+   FC_ASSERT( !current_bitasset_data.has_settlement(), "Cannot update a bitasset after a global settlement has executed" );
+
+   bool after_hf_core_922_931 = ( d.get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_922_931_TIME );
+
+   // Are we changing the backing asset?
+   if( op.new_options.short_backing_asset != current_bitasset_data.options.short_backing_asset )
    {
-      FC_ASSERT(a.dynamic_asset_data_id(d).current_supply == 0);
-      FC_ASSERT(d.find_object(o.new_options.short_backing_asset));
+      FC_ASSERT( asset_obj.dynamic_asset_data_id(d).current_supply == 0,
+                 "Cannot update a bitasset if there is already a current supply." );
+
+      const asset_object& new_backing_asset = op.new_options.short_backing_asset(d); // check if the asset exists
 
-      if( a.issuer == GRAPHENE_COMMITTEE_ACCOUNT )
+      if( after_hf_core_922_931 ) // TODO remove this check after hard fork if things in `else` did not occur
       {
-         const asset_object& backing = a.bitasset_data(d).options.short_backing_asset(d);
-         if( backing.is_market_issued() )
+         FC_ASSERT( op.new_options.short_backing_asset != asset_obj.get_id(),
```

Comment:
IMO this is not enough, we should also protect against cycles.

Reply:
Actually I want to give up on this. A cycle can be long (more discussion will be in another comment).

Bottom line, it's unable to change backing asset when there is a supply, IMHO even if there are cycles, the consequence won't be too serious.

Reply:
Zero supply is a precondition only for the asset to be changed, not for the backing asset. If the backing asset has non-zero supply it will be possible to borrow.

Reply:
Also, cycles shouldn't be long due to the other rule (MPA backed by MPA backed by MPA is forbidden).

Reply:
> Zero supply is a precondition only for the asset to be changed, not for the backing asset. If the backing asset has non-zero supply it will be possible to borrow.

Changing backing asset of "the asset to be changed" won't create a cycle, only changing the backing asset of "the backing asset" will. So if there is a cycle, it means all assets in the cycle has zero supply, in this case, no supply can be made since need to put something as collateral.